system/Security/Security.php

```diff
@@ -206,12 +216,14 @@ public function CSRFVerify(RequestInterface $request)
 		{
 			return $this->CSRFSetCookie($request);
 		}
-
-		// Do the token exist in _POST or php://input (json) data?
-		$CSRFTokenValue = $_POST[$this->CSRFTokenName] ??
-            (!empty($input = file_get_contents('php://input')) && !empty($json = json_decode($input)) && json_last_error() === JSON_ERROR_NONE ?
-                ($json->{$this->CSRFTokenName} ?? null) :
-                null);
+		
+		// Do the tokens exist in _POST, HEADER or optionally php:://input - json data
+        $CSRFTokenValue = $_POST[$this->CSRFTokenName] ??
+            (!is_null($request->getHeader($this->CSRFHeaderName)) && !empty($request->getHeader($this->CSRFHeaderName)->getValue()) ?
+                $request->getHeader($this->CSRFHeaderName)->getValue() :
+                (!empty(file_get_contents('php://input')) && !empty($json = json_decode(file_get_contents('php://input'))) && json_last_error() === JSON_ERROR_NONE ?
```

Comment:
```suggestion
                (!empty($request->getBody()) && !empty($json = json_decode($request->getBody())) && json_last_error() === JSON_ERROR_NONE ?
```
With this change this code become testable. We can't mock `php://input` - also I think it's a proper way to do it.